src/jit.c

```diff
@@ -0,0 +1,1578 @@
+#endif
+    state->offset += len;
+}
+
+#if defined(__x86_64__)
+static inline void emit1(struct jit_state *state, uint8_t x)
+{
+    emit_bytes(state, &x, sizeof(x));
+}
+
+static inline void emit2(struct jit_state *state, uint16_t x)
+{
+    emit_bytes(state, &x, sizeof(x));
+}
+
+static inline void emit4(struct jit_state *state, uint32_t x)
+{
+    emit_bytes(state, &x, sizeof(x));
+}
+
+static inline void emit8(struct jit_state *state, uint64_t x)
+{
+    emit_bytes(state, &x, sizeof(x));
+}
+
+static inline void emit_modrm(struct jit_state *state, int mod, int r, int m)
+{
+    assert(!(mod & ~0xc0));
+    emit1(state, (mod & 0xc0) | ((r & 7) << 3) | (m & 7));
+}
+
+static inline void emit_modrm_reg2reg(struct jit_state *state, int r, int m)
+{
+    emit_modrm(state, 0xc0, r, m);
+}
+
+static inline void emit_modrm_and_displacement(struct jit_state *state,
+                                               int r,
+                                               int m,
+                                               int32_t d)
+{
+    if (d == 0 && (m & 7) != RBP) {
+        emit_modrm(state, 0x00, r, m);
+    } else if (d >= -128 && d <= 127) {
+        emit_modrm(state, 0x40, r, m);
+        emit1(state, d);
+    } else {
+        emit_modrm(state, 0x80, r, m);
+        emit4(state, d);
+    }
+}
+
+static inline void emit_rex(struct jit_state *state, int w, int r, int x, int b)
+{
+    assert(!(w & ~1));
+    assert(!(r & ~1));
+    assert(!(x & ~1));
+    assert(!(b & ~1));
+    emit1(state, 0x40 | (w << 3) | (r << 2) | (x << 1) | b);
+}
+
+/* Emit a REX prefix incorporating the top bit of both src and dst. This step is
+ * skipped if no bits are set.
+ */
+static inline void emit_basic_rex(struct jit_state *state,
+                                  int w,
+                                  int src,
+                                  int dst)
+{
+    if (w || (src & 8) || (dst & 8))
+        emit_rex(state, w, !!(src & 8), 0, !!(dst & 8));
+}
+
+static inline void emit_push(struct jit_state *state, int r)
+{
+    emit_basic_rex(state, 0, 0, r);
+    emit1(state, 0x50 | (r & 7));
+}
+
+static inline void emit_pop(struct jit_state *state, int r)
+{
+    emit_basic_rex(state, 0, 0, r);
+    emit1(state, 0x58 | (r & 7));
+}
+
+static inline void emit_jump_target_address(struct jit_state *state,
+                                            int32_t target_pc)
+{
+    struct jump *jump = &state->jumps[state->n_jumps++];
+    jump->offset_loc = state->offset;
+    jump->target_pc = target_pc;
+    emit4(state, 0);
+}
+#elif defined(__aarch64__)
+static void emit_instruction(struct jit_state *state, uint32_t instr)
+{
+    assert(instr != BAD_OPCODE);
+    emit_bytes(state, &instr, 4);
+}
+
+/* Get the value of the size bit in most instruction encodings (bit 31). */
+static uint32_t sz(bool is64)
+{
+    return (is64 ? UINT32_C(1) : UINT32_C(0)) << 31;
+}
+
+
+/* For details on ARM instructions, users can refer to
```

Comment:
You should refer it to "Arm instructions" instead of "ARM instructions."
See https://www.arm.com/company/policies/trademarks/guidelines-trademarks
> Do not alter the spelling or form of Arm's trademarks by abbreviating them, creating acronyms, translating them, joining them to other words, symbols or numbers (either as one word or with a hyphen - unless otherwise permitted, e.g. Arm® Cortex®-A15), or using improper capitalization. However, permitted capitalization occurs when using an Arm word trademark in headlines, titles or text where all of the surrounding words are shown in uppercase characters. In this situation, you may use the relevant Arm trademark in uppercase characters, provided that such use complies with these guidelines.